lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -1315,17 +1315,19 @@ private function getSQLForJoins(string $fromAlias, array &$knownAliases) : strin
 
         if (isset($this->sqlParts['join'][$fromAlias])) {
             foreach ($this->sqlParts['join'][$fromAlias] as $join) {
-                if (array_key_exists($join['joinAlias'], $knownAliases)) {
-                    throw NonUniqueAlias::new($join['joinAlias'], array_keys($knownAliases));
+                /** @var QueryPartJoin $join */
+                if (array_key_exists($join->joinAlias, $knownAliases)) {
+                    throw NonUniqueAlias::new($join->joinAlias, array_keys($knownAliases));
                 }
-                $sql                             .= ' ' . strtoupper($join['joinType'])
-                    . ' JOIN ' . $join['joinTable'] . ' ' . $join['joinAlias']
-                    . ' ON ' . ((string) $join['joinCondition']);
-                $knownAliases[$join['joinAlias']] = true;
+                $sql                           .= ' ' . strtoupper($join->joinType)
+                    . ' JOIN ' . $join->joinTable . ' ' . $join->joinAlias
+                    . ' ON ' . ((string) $join->joinCondition);
```

Comment:
Same comment as in #3829:

> Strangely, AFAICS in the source code (and confirmed by no zero broken tests), `joinCondition` is `string|null`. But casting `null` to `string` here would result in `ON` nothing, which would be a syntax error. I'm not sure what the intention was here.

Reply:
I think we should replace casting with a condition. A `JOIN` without `ON` is syntactically valid but an empty `ON` isn't. It doesn't have to be done now since we'd need a test for it. Mind filing an issue?

Reply:
The last sticking point is this `ON (empty string)`. Should I change it to only add the `ON` part if the condition is not null? **Was this really a bug that went unnoticed all that time?**

Reply:
I don't mind either doing it in this PR, or, if you want to keep changes self-contained, you can merge this first and I'll create another PR with the fix and a test.

Reply:
> Should I change it to only add the `ON` part if the condition is not null?

Not here but yes, in a separate PR that would have a test that reproduces the actual issue.

> Was this really a bug that went unnoticed all that time?

It looks like that. This logic exists since the initial implementation of the query builder (https://github.com/doctrine/dbal/commit/ae833d153f77a44c3737e7760bebfc992b579cd5).

Reply:
> I'll create another PR with the fix and a test.

Once the test confirms that it's a but, the other PR may target `2.10` instead of `master` so that we can release the fix sooner.

Reply:
Sure, but I'll have to create 2 PRs I guess, as 2.10 will need the fix with the associative array code, and master will need the fix with the Join class?